Changes necessary to support `netctl login`

dseevr opened the pull request:
This commit does not break any existing functionality for regular netmaster communication

Changes:

- default http client is replaced by a http.Client in the ContivClient struct
- httpGet()/httpPost()/httpDelete() are moved into ContivClient and use the new internal http.Client
- added SetHttpClient() which allows for overriding the internal http.Client (to disable cert checking, add timeouts, etc.)
- added Login() function which performs a login POST request against auth_proxy
- added SetAuthToken() which adds a custom request header (X-Auth-Token) before requests are sent out
- httpGet()/httpPost()/httpDelete() are modified to set any custom request headers before performing the request

Signed-off-by: Bill Robinson <dseevr@users.noreply.github.com>

jojimt commented on generators/templates.go:
```diff
@@ -87,14 +87,70 @@ package client
+
+type NetctlConfig struct {
+	Token string ` + "`" + `json:"token"` + "`" + `
+}
```
This is a generic go-client for contiv. We should avoid explicit references to netctl here.

dseevr replied:
Makes sense to me!

This was a quick hack to see if I could add `netctl login` functionality in a reasonable amount of time without breaking anything.  It works, now it just needs to be done in an acceptable manner.  Any architecture suggestions you have would be valuable

jojimt commented on generators/templates.go:
```diff
@@ -87,14 +87,70 @@ package client
+		return err
+	}
+
+	netctlConfigFile := usr.HomeDir + "/.netctl/config.json"
```
Is it more appropriate to make the config file part of netctl rather than the model and have netctl pass it to the client?

rhim replied:
So, instead of having netctl specific implict file, maybe AuthTokenHeader should just add in a generic token variable value, and netctl as client would manage that token variable by itself (thaw it to file, read it from file)

dseevr replied:
Yup.  Will make the requested changes today.  Thanks 👍

dseevr committed:
Changes necessary to support `netctl login`

- default http client is replaced by a http.Client in the ContivClient struct
- httpGet()/httpPost()/httpDelete() are moved into ContivClient and use the new internal http.Client
- added SetHttpClient() which allows for overriding the internal http.Client (to disable cert checking, add timeouts, etc.)
- added Login() function which performs a login POST request against auth_proxy
- added SetAuthToken() which adds a custom request header (X-Auth-Token) before requests are sent out
- httpGet()/httpPost()/httpDelete() are modified to set any custom request headers before performing the request

Signed-off-by: Bill Robinson <dseevr@users.noreply.github.com>

dseevr commented:
@jojimt @rhim PTAL

I completely overhauled how the code is working.  Please take a look at the PR description for the updated changes overview before looking at the code

I also updated https://github.com/contiv/netplugin/pull/823.  Same deal, description is updated with high-level changes

This is much more sane now 😄

jojimt commented:
Yes, design-wise this looks good. Can we add some test cases to cover this as well? It is also easier to review the changes that way.

rhim reviewed (commented):
could you please clarify why there is duplication between templates.go and clientHdr.tmpl?

dseevr commented:
@rhim templates.go is generated from the .tmpl files when you run `make`.  After you commit it, you run `make` in contivmodel to bring in the changes from this repo

The diffs aren't quite identical and you can end up with some weird-looking code (see: `loginPayload` struct's JSON in templates.go), but it ends up looking normal when imported into contivmodel (compare templates.go's weird JSON to contivModelClient.go in that repo's PR)
